### PR TITLE
WIP: add WebGL debugMode() 

### PR DESCRIPTION
**Pull request by AidanNelson**
This PR adds a `debugMode()` function to the webGL mode.  The function works to make visualizing 3D space somewhat easier by adding a grid 'ground' plane and an axes icon indicating the +X, +Y, and +Z directions.  Also...

* adds Interaction submodule to webGL YUI doc organization to contain debugMode and orbitControl
* adds `_grid()` and `_axes()` private functions to be added to preregisteredMethods by `debugMode()`.

If there is a cleaner way to overload this function, please let me know.

### CODE REVIEW

File: test/manual-test-examples/webgl/interaction/debugMode/sketch.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+function setup() {
+  createCanvas(windowWidth, windowHeight, WEBGL);
+  normalMaterial();
+  let cam = createCamera();
+  cam.move(0, -this.height / 4, 0);
+  debugMode();
+}
```
Comment (kjhollen) [inline test/manual-test-examples/webgl/interaction/debugMode/sketch.js]: but also: I can't run this because it generates an error with validateParams in my default build...

File: src/webgl/interaction.js [hunk 2]
```diff
@@ -106,4 +114,369 @@ p5.prototype.orbitControl = function(sensitivityX, sensitivityY) {
...
+ *   debugMode();
+ * }
+ *
+ * function draw() {
+ *   background(200);
+ *   orbitControl();
+ *   box(15, 30);
+ *   // Press the spacebar to turn debugMode off!
+ *   if (keyIsDown(32)) {
+ *     debugMode(OFF);
```
[the remark this answers is not in the document]
Comment (AidanNelson) [reply src/webgl/interaction.js]: Yes, that is much better!

File: src/webgl/interaction.js [hunk 3]
```diff
@@ -106,4 +114,369 @@ p5.prototype.orbitControl = function(sensitivityX, sensitivityY) {
...
+ *   debugMode(100, 10, 0, 0, 0, 20, 0, -40, 0);
+ * }
+ *
+ * function draw() {
+ *   background(200);
+ *   orbitControl();
+ *   box(15, 30);
+ * }
+ * </code>
+ * </div>
```
[the remark this answers is not in the document]
Comment (AidanNelson) [reply src/webgl/interaction.js]: yea, this is because #3063 broke orbitControl (which was still using the old camera paradigm) and I didn't realize until after it was merged 😳.  

#3088 will fix orbitControl so that it works with p5.Camera object.

File: src/webgl/interaction.js [hunk 4]
```diff
@@ -106,4 +114,369 @@ p5.prototype.orbitControl = function(sensitivityX, sensitivityY) {
...
+ */
+
+/**
+ * @method debugMode
+ * @param {Constant} mode GRID, AXES, or OFF
```
Comment (Spongman) [inline src/webgl/interaction.js]: this should be
```
 * @param {Constant} mode either GRID, AXES, or OFF
```

File: test/manual-test-examples/webgl/interaction/debugMode/sketch.js (hunk 1 above)
Comment (Spongman) [reply test/manual-test-examples/webgl/interaction/debugMode/sketch.js]: there's a bug in the validateParameters handling that causes a crash when the description of a {Constant}-typed parameter does not match the required format.
the fix in this case is to add the 'either' keyword before the list of valid values for the parameter ("**either** GRID, AXES, or OFF").

File: src/webgl/interaction.js [hunk 5]
```diff
@@ -106,4 +114,369 @@ p5.prototype.orbitControl = function(sensitivityX, sensitivityY) {
...
+ * default grid and axes icon, or it can be called according to the examples
+ * above to customize the size and position of the grid and/or axes icon.
+ *
+ * By default, the grid will run through the origin (0,0,0) of the sketch
+ * along the XZ plane
+ * and the axes icon will be offset from the origin.  Both the grid and axes
+ * icon will be sized according to the current canvas size.  Note that because the
+ * grid runs parallel to the default camera view, it is often helpful to use
+ * debugMode along with orbitControl to allow full view of the grid.
+ * @method debugMode
```
[the remark this answers is not in the document]
Comment (AidanNelson) [reply src/webgl/interaction.js]: This seems to have been fixed by adding 'either' after the `@param {Constant} mode` tag, per @Spongman's suggestion below.

File: test/manual-test-examples/webgl/interaction/debugMode/sketch.js (hunk 1 above)
Comment (Spongman) [reply test/manual-test-examples/webgl/interaction/debugMode/sketch.js]: indeed, going to add that now...

Comment (Spongman) [reply test/manual-test-examples/webgl/interaction/debugMode/sketch.js]: sorry, i should have done this a long time ago... #3122

File: src/webgl/interaction.js [hunk 6]
```diff
@@ -106,4 +114,369 @@ p5.prototype.orbitControl = function(sensitivityX, sensitivityY) {
...
+ *   background(200);
+ *   orbitControl();
+ *   box(15, 30);
+ *   // Press the spacebar to turn debugMode off!
+ *   if (keyIsDown(32)) {
+ *     debugMode(OFF);
+ *   }
+ * }
+ * </code>
+ * </div>
```
[the remark this answers is not in the document]
Comment (AidanNelson) [reply src/webgl/interaction.js]: added!

File: src/webgl/interaction.js [hunk 7]
```diff
@@ -106,4 +114,369 @@ p5.prototype.orbitControl = function(sensitivityX, sensitivityY) {
...
+      this._renderer._curCamera.cameraMatrix.mat4[12],
+      this._renderer._curCamera.cameraMatrix.mat4[13],
+      this._renderer._curCamera.cameraMatrix.mat4[14],
+      this._renderer._curCamera.cameraMatrix.mat4[15]
+    );
+
+    // Lines along X axis
+    for (var q = 0; q <= numDivs; q++) {
+      this.strokeWeight(1);
+      this.stroke(0, 0, 0);
```
[the remark this answers is not in the document]
Comment (AidanNelson) [reply src/webgl/interaction.js]: Good point, that would be nice!  

I've set it to use the last-set stroke color and stroke weight.  This introduces an odd behavior when called without first setting a stroke, where it will use the stroke color last set by `_axesIcon()`, but I suspect this is fixed by #3076.

**Comment by AidanNelson** [comment]
@kjhollen these most recent changes also simplify the argument-checking.  Let me know if you prefer the previous version.  

And sorry about the number of commits!  Let me know if you'd like me to rebase them...

File: src/webgl/interaction.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (kjhollen) [reply src/webgl/interaction.js]: woops, okay, it looks like I can't fully test this until we get 3088 in—I think it's ready to go, but just had one more quick question there. I'll take a closer look at this and the _orbit tests tomorrow morning!

[the remark this answers is not in the document]
Comment (kjhollen) [reply src/webgl/interaction.js]: @AidanNelson I just merged #3088, do you want to go ahead and merge this branch with webgl-gsoc-2018 and push again?

[the remark this answers is not in the document]
Comment (kjhollen) [reply src/webgl/interaction.js]: cool, confirming that this is resolved with a merge of the latest webgl-gsoc-2018 branch!

File: src/webgl/interaction.js [hunk 8]
```diff
@@ -140,4 +143,423 @@ p5.prototype.orbitControl = function(sensitivityX, sensitivityY) {
   return this;
 };
 
+/**
+ * DebugMode() helps visualize 3D space by adding a grid to indicate where the
```
Comment (kjhollen) [inline src/webgl/interaction.js]: use lowercase 'd' here, sorry I didn't catch this one before!

File: src/webgl/interaction.js [hunk 9]
```diff
@@ -140,4 +143,423 @@ p5.prototype.orbitControl = function(sensitivityX, sensitivityY) {
...
+ *   box(15, 30);
+ *   // Press the spacebar to turn debugMode off!
+ *   if (keyIsDown(32)) {
+ *     noDebugMode();
+ *   }
+ * }
+ * </code>
+ * </div>
+ * @alt
+ * a 3D box is centered on a grid in a 3D sketch. there is also an icon
```
Comment (kjhollen) [inline src/webgl/interaction.js]: can you summarize what the icon is concisely? three lines + color + direction?
